src/attention.jl

```diff
@@ -0,0 +1,139 @@
+
+# Examples
+    
+```julia
+q, k, v = rand(10, 20, 2), rand(10, 30, 2), rand(20, 30, 2)
+y, α = dot_product_attention(q, k, v)
+```
+"""
+function dot_product_attention(q::AA{N}, k::AA{N}, v::AA{N}, args...; kws...) where N
+    batch_size = size(q)[3:end]
+    batch_size == size(k)[3:end] == size(v)[3:end] || throw(ArgumentError("Batch dimensions have to be the same."))
+    q, k, v = map(x -> reshape(x, size(x, 1), size(x, 2), :), (q, k, v))
+
+    x, α = dot_product_attention(q, k, v, args...; kws...)
+
+    x = reshape(x, size(x, 1), size(x, 2), batch_size...)
+    α = reshape(α, size(α)[1:3]..., batch_size...)
+    return x, α
+end
+
+function dot_product_attention(q::AA3, k::AA3, v::AA3, bias=nothing; 
+            fdrop=identity, mask=nothing, nheads=1)
+
+    (size(q, 3) == size(k, 3) == size(v, 3)) || throw(ArgumentError("Batch dimensions have to be the same."))
+    size(q, 1) == size(k, 1) || throw(ArgumentError("First dimension in query and key has to be the same."))
+    size(k, 2) == size(v, 2) || throw(ArgumentError("Second dimension in key and value has to be the same."))
+    
+    # Multihead attention. TODO create fastpath for singlehead attention.
+    q, k, v = split_heads.((q, k, v), nheads)
+    x, α = _dot_product_attention(q, k, v, bias, fdrop, mask)
+    return join_heads(x), α
+end
+
+function _dot_product_attention(q::AA4, k::AA4, v::AA4, bias, fdrop, mask)
+    # [q] = [qk_dim ÷ nheads, nheads, q_len, batch_size]
+    # [k] = [qk_dim ÷ nheads, nheads, kv_len, batch_size]
+    # [v] = [v_dim ÷ nheads, nheads, kv_len, batch_size]
+
+    α = dot_product_attention_scores(q, k, bias; fdrop, mask)
+    # [α] = [kv_len, q_len, nheads, batch_size]
+    
+    # The following permutedims and batched_mul are equivalent to
+    # @tullio x[d, h, i, b] := α[j, i, h, b] * v[d, h, j, b]
+    vt = permutedims(v, (1, 3, 2, 4))
+    x = batched_mul(vt, α)
+    x = permutedims(x, (1, 3, 2, 4))
+    # [x] = [kv_dim ÷ nheads, nheads, q_len, batch_size]
+    return x, α
+end
+
+"""
+    dot_product_attention_scores(query, key, [bias]; [fdrop, mask])
+
+Return the attention scores for the [`dot_product_attention`](@ref).
+Input arrays must have dimensions 
+`(num_features ÷ nheads, nheads, sequence_length, batch_size)`.
+
+See [`dot_product_attention`](@ref) for more details.
+"""
+function dot_product_attention_scores(q::AA4{T}, k::AA4{T}, bias=nothing; 
+            fdrop=identity, mask=nothing) where T
+
+    # The following permutedims and batched_mul are equivalent to
+    # @tullio logits[j, i, h, b] := q[d, h, i, b] * k[d, h, j, b] / √T(qk_dim)
+    kt = permutedims(k, (3, 1, 2, 4))
+    qt = permutedims(q, (1, 3, 2, 4)) ./ √T(size(q, 1))
+    logits = batched_mul(kt, qt)
+    # [logits] = [kv_len, q_len, nheads, batch_size]
+
+    if bias !== nothing
+        logits = logits .+ bias
+    end
+
+    if mask !== nothing
+        if mask === :causal
+            mask = make_causal_mask(logits)
+        end
```

Comment:
I think a cleaner API would be to let the `mask` keyword be a function. The `nothing` case is `mask = identity` and the causal case is `mask = make_causal_mask` (which I feel should be just `causal_mask` to be succinct).

Is there a reason to construct the mask on the fly? The calling layer in Flux can probably make and store the mask once. Then the other option is to allow `nothing` or an array. Then the user passes in `mask = causal_mask(ntoken)`.

Reply:
You're right, it shouldn't be `identity`, it should be `trues_like` though I'd be okay with `nothing` in order skip computing a mask at all.

My comment about constructing on the fly was not a performance concern. I just think it is more intuitive to pass in exactly the mask array I want used. It's an easier rule to remember and also scalable to whatever masking scheme is desired.

Reply:
The downside is that you have to make an array the right size. If you have several layers and want the same scheme for each, then perhaps it's a pain. Whereas a function like `trues_like` is told the size automatically.

Reply:
That's true, but generally the size of this matrix which is `# of tokens X # of tokens` is known ahead of time. Even so, I agree that not needing to pass in this info is cleaner.

I mostly wanted to avoid "symbol switches" for arguments.

Reply:
Yes to avoiding symbols. I like I like this `mask = trues_like` proposal the best so far.

One question I haven't looked at is what format the CUDNN thing is going to want.

Reply:
Instead of saying `mask` is either an array or callable, could we say it should be either an array or marker type for which one can override some `init_mask(x, k, v)` function? This would allow us to shift the conditionals out of the attention functions, while still allowing for relatively terse syntax like `mask = CausalMask()` when users don't want to precompute their own. You could imagine nice party tricks like passing `mask = I`.

Reply:
https://github.com/FluxML/NNlib.jl/pull/460 is a go at this masked softmax idea. 

With that, the default of no mask can in fact be `mask = Returns(true)` here, instead of `trues_like`. And the terse causal mask can be `const causal_mask = triu ∘ trues_like`, or a function equivalent to this (maybe it can be more efficient, not sure `triu`  works on CuArrays). No conditionals required.

Reply:
`triu` only works on `AbstractMatrix`, which is not sufficient for the attention.

Reply:
For this first implementation, I prefer to keep it more minimalistic and just accept `nothing` or arrays (I will remove `:causal`)